### PR TITLE
Game action fixes

### PR DESCRIPTION
**Pull request by ZehMatt**
Fix for #6755 and #6965

### CODE REVIEW

File: src/openrct2/actions/GameActionCompat.cpp [hunk 1]
```diff
@@ -237,4 +238,43 @@ extern "C"
...
+        if (!(flags & GAME_COMMAND_FLAG_APPLY))
+        {
+            auto res = GameActions::Query(&gameAction);
+            if (res->Error == GA_ERROR::OK)
+                cost = res->Cost;
+        }
+        else
+        {
+            auto res = GameActions::Execute(&gameAction);
+            if(res->Error == GA_ERROR::OK)
```
Comment (duncanspumpkin) [inline src/openrct2/actions/GameActionCompat.cpp]: Common code

**Comment by duncanspumpkin** [comment]
Can you check that there aren't error titles for this game command.

**Comment by ZehMatt** [comment]
I don't quite understand, which error titles?

**Comment by duncanspumpkin** [comment]
Error titles tend to be defined not in the game command. You need to check how the existing error is displayed as you might be removing the title. Have a look at the changes I made to the other game actions

**Comment by ZehMatt** [comment]
Gotcha, I'll have a look.

**Comment by ZehMatt** [comment]
I removed all the error titles and spotted an issue where the global gGameCommandErrorText has to be assigned as its used in ride_construction_tooldown_construct for its trial and error runs. So I decided to assign them in maze_set_track compatibility function to keep the behavior. This can be refactored once we have no more C code left.

**Comment by duncanspumpkin** [comment]
You've removed the error title now but you haven't put it in the game action (as far as i can tell from my quick look)

File: src/openrct2/actions/MazeSetTrackAction.hpp [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        auto res = std::make_unique<GameActionResult>();
+
+        res->Position.x = _x + 8;
+        res->Position.y = _y + 8;
+        res->Position.z = _z;
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_RIDE_CONSTRUCTION;
```
Comment (duncanspumpkin) [inline src/openrct2/actions/MazeSetTrackAction.hpp]: Res->errortitle should be set here

Comment (ZehMatt) [reply src/openrct2/actions/MazeSetTrackAction.hpp]: Fixed, now it shows up again.

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 3]
```diff
@@ -3549,15 +3549,13 @@ void ride_construction_tooldown_construct(sint32 screenX, sint32 screenY)
...
-                _currentTrackBeginX,
-                GAME_COMMAND_FLAG_APPLY | (4 << 8),
+            _trackPlaceCost = maze_set_track(_currentTrackBeginX,
                 _currentTrackBeginY,
-                _currentRideIndex,
-                GAME_COMMAND_SET_MAZE_TRACK,
                 _currentTrackBeginZ,
-                0);
+                GAME_COMMAND_FLAG_APPLY,
+                4,
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/RideConstruction.cpp]: is 4 a direction?

File: src/openrct2/actions/RideDemolishAction.hpp [hunk 4]
```diff
@@ -217,4 +218,93 @@ struct RideDemolishAction : public GameActionBase<GAME_COMMAND_DEMOLISH_RIDE, Ga
...
+        if (queryRes->Error == GA_ERROR::OK)
+        {
+            auto execRes = setMazeTrack.Execute();
+            if (execRes->Error == GA_ERROR::OK)
+            {
+                return execRes->Cost;
+            }
+        }
+
+        return MONEY16_UNDEFINED;
```
Comment (duncanspumpkin) [inline src/openrct2/actions/RideDemolishAction.hpp]: MONEY32_UNDEFINED

**Comment by duncanspumpkin** [comment]
It looks to me like 4 as a direction is being used as an additional flag to the function. I would change it to an explicit bool variable to improve code readability.

File: src/openrct2/actions/RideDemolishAction.hpp (hunk 4 above)
Comment (ZehMatt) [reply src/openrct2/actions/RideDemolishAction.hpp]: God damnit, why are there two, I keep messing this up.

**Comment by ZehMatt** [comment]
If I would to introduce a bool then what would be the default direction value? That makes things even harder to read. Also I'm not sure what I should name it.

**Comment by duncanspumpkin** [comment]
I've not got my head fully into this bit of code but from my quick review the direction is being used thus:

1. If direction is 4 then only place the main maze element don't remove hedge (This is used for initial placement of the maze)

2. If direction is not 4 then it is in the range 0-3 and the direction tells the function where to remove hedge from to make it join up correctly.

To me it looks like you should pass a bool for initial placement and when set it ignores the direction parameter and performs 1. When not set it performs 2.

**Comment by ZehMatt** [comment]
Alrite, I'll call it initialPlacement then.

**Pull request merged**